src/Razor/src/Microsoft.VisualStudio.RazorExtension/SyntaxVisualizer/SyntaxVisualizerControl.xaml.cs

```diff
@@ -181,13 +251,82 @@ public void ShowGeneratedHtml()
             return;
         }
 
+        EnsureInitialized();
+
+        if (_languageServerFeatureOptions.UseRazorCohostServer)
+        {
+            if (_fileUriProvider.TryGet(_activeWpfTextView.TextBuffer, out var hostDocumentUri))
+            {
+                var request = new DocumentContentsRequest
+                {
+                    TextDocument = new TextDocumentIdentifier { DocumentUri = hostDocumentUri },
+                    Kind = GeneratedDocumentKind.Html
+                };
+
+                var response = _joinableTaskFactory.Run(async () =>
+                {
+                    var lspResponse = await _lspRequestInvoker.ReinvokeRequestOnServerAsync<DocumentContentsRequest, DocumentContentsResponse>(
+                        _activeWpfTextView.TextBuffer,
+                        "razor/generatedDocumentContents",
+                        RazorLSPConstants.RoslynLanguageServerName,
+                        request,
+                        CancellationToken.None);
+
+                    return lspResponse?.Response;
+                });
+
+                if (response != null)
+                {
+                    OpenGeneratedCode(response.FilePath, response.Contents);
+                    return;
+                }
+            }
+        }
+
+        // Fall back to legacy method if cohosting is not enabled or failed
         OpenVirtualDocuments<HtmlVirtualDocumentSnapshot>(_activeWpfTextView.TextBuffer);
     }
 
     public void ShowSerializedTagHelpers(TagHelperDisplayMode displayKind)
     {
         ThreadHelper.ThrowIfNotOnUIThread();
 
+        EnsureInitialized();
+
+        if (_languageServerFeatureOptions.UseRazorCohostServer && _activeWpfTextView is not null)
+        {
+            if (_fileUriProvider.TryGet(_activeWpfTextView.TextBuffer, out var hostDocumentUri))
+            {
+                var response = _joinableTaskFactory.Run(async () =>
+                {
+                    var lspResponse = await _lspRequestInvoker.ReinvokeRequestOnServerAsync<TextDocumentIdentifier, string>(
```

Comment:
Looks like I forgot about TagHelperDisplayMode. That means we need to change the endpoint for tag helpers to take a new TagHelpersRequest object, with a TextDocumentIdentifier and a new TagHelpersKind enum, which mirrors the TagHelperDisplayMode enum. Adjust the test and service definitions and implementations as necessary, copying the logic from the switch statement below in the remote service implementation. Make sure to create a static Create method on the new TagHelpersRequest object, like with DocumentContentsRequest

Probably best to action the other feedback first, then do this one in a separate commit.